### PR TITLE
API: deprecate show_in_notebook

### PR DESCRIPTION
**Pull request by MridulS**
### Description

Let's take a bit agressive way to fix https://github.com/astropy/astropy/issues/15566, this PR proposes to deprecate `show_in_notebook()` and all notebook (classic) specific JS code. The Jupyter Notebook has moved to using the JupyterLab plumbing (from v7) and the classic notebook is on [maintainence mode](https://github.com/jupyter/notebook#maintained-versions). Things will probably move fast in the JupyterLab world and the JS specific code in astropy will only work with classic notebook. I would argue that astropy shouldn't be in the business of JS specific logic, this should either be delagated to other third party tools like https://github.com/bloomberg/ipydatagrid, https://github.com/mwouts/itables or the jupyter specific code be moved to a new package something like `astropy-jupyter-extension`. This should make the maintenance much more sustainable.

There is some conversation on Slack regarding this: https://astropy.slack.com/archives/C067V74GK/p1705405195253679

Well this "fixes" a lot of `show_in_notebook` issues 😅 

cc @hamogu @pllim @mwcraig @dhomeier 

fix https://github.com/astropy/astropy/issues/11836, fix https://github.com/astropy/astropy/issues/9660 (partially, as this feature could still be added to show_in_browser), fix https://github.com/astropy/astropy/issues/15566, fix https://github.com/astropy/astropy/issues/10317

### CODE REVIEW

**Comment by pllim** [comment]
I'll defer to subpackage maintainers and Jupyter experts here. The test fails because you need to handle this new warning. Also would need a change log.

Thanks!

File: astropy/table/table.py
Comment (saimn) [inline astropy/table/table.py:1839]: `show_in_browser` is not related to Jupyter, and is not broken. It creates a standalone HTML file and was the original reason of using datatables etc., the notebook integration came later. 
Not sure if it's widely used though.

Comment (MridulS) [reply astropy/table/table.py:1839]: True, but if we are going to remove all JS code, it *might* make sense to remove all the related bits. It should also help with resolving issues like https://github.com/astropy/astropy/issues/10420 and https://github.com/astropy/astropy/issues/11188

I know I'm being a bit bold here and marking it as a pending deprecation may tell us if there are people out there using this bit :D

Comment (mwcraig) [reply astropy/table/table.py:1839]: I agree `show_in_browser` is not Jupyter-related. What about splitting this into two PRs? I think deprecating `show_in_notebook` is the right call, and that we shouldn't replace it. The underpinnings of notebook change too often for us to support well in astropy. 

I'm less clear about `show_in_browser`. 

On the one hand, it works. On the other hand, it adds dependencies that we are bundling in `jsviewer.py` that are not update regularly by dependabot. If we are going to supply `show_in_browser` then it seems worth discussing separately from `show_in_notebook`.

At the end of the day, I lean towards deprecating `show_in_browser` too and seeing what the reaction from the community is...

**Review by mwcraig — COMMENTED** [review]
See my inline comments. To summarize, I'd recommend changing this notebook to deprecate `show_in_notebook` and a separate PR to handle `show_in_browser`.

**Comment by hamogu** [comment]
Similarly, one might also use a [bokeh `DataTable`](https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#datatable). 

My point is that instead of offloading the task of displaying the html table to a js library that we bundle, we could do that indirectly and offload to some other python package from the scientific ecosystem where js support is closer to their core mission and have them worry about which version to bundle and howto handle notebook updates etc.

I think of this a mostly a documentation task: "If you want to display your `astropy.table.Table` nicely, here are a few packages that can do that for you and here is a simple example how your astropy table can be passed to them and displayed ... "

**Comment by MridulS** [comment]
Thanks all for your reviews!

Yes I think it probably makes sense to link to other well maintained libraries for working with Tables interactively inside a jupyter environment. I have removed the deprecation of `JSViewer` in this PR, I'll open a new one so we can discuss this it there.

**Review by mwcraig — APPROVED** [review]
In the short run we can just open an issue to note that it would be nice to recommend a replacement rather than hashing that out here.

File: astropy/table/table.py
Comment (pllim) [inline astropy/table/table.py:1760]: Was there a conversation on why you opt to go with PendingDeprecationWarning instead of a full DeprecationWarning? This does not officially kick off the deprecation period AFAIK.

Also change log is missing.

**Review by pllim — CHANGES_REQUESTED** [review]
Thanks, but I have a question and a comment.

File: astropy/table/table.py
Comment (MridulS) [reply astropy/table/table.py:1760]: Nope, not yet. I just put in a `PendingDeprecationWarning` as I wasn't sure if there is enough consensus in deprecating it. I can change this now with the change log entry.

Comment (taldcroft) [reply astropy/table/table.py:1760]: It seems the consensus is to go ahead with deprecation now on both methods.

**Comment by eerovaher** [comment]
There are good reasons to deprecate the functions in question, but I would like there to be documentation about available alternatives, and if we do write such documentation then the deprecation message should point to it.

**Comment by taldcroft** [comment]
> There are good reasons to deprecate the functions in question, but I would like there to be documentation about available alternatives, and if we do write such documentation then the deprecation message should point to it.

I'm not aware of any easy alternatives to `show_in_notebook()`. The `pandas` styling stuff is basically equivalent to the table repr in a notebook but AFAIK it does not have any javascript bling like showing only 50 rows or sorting by columns.

**Comment by pllim** [comment]
tl;dr -- I see one approval, so is the diff okay as-is or more discussion needed? If so, what?

File: docs/changes/table/15905.api.rst
Comment (pllim) [inline docs/changes/table/15905.api.rst:1]: Please avoid "live linking" in change log entry. Thanks!

```suggestion
``show_in_notebook`` is deprecated and it is recommended to use dedicated
```

**Comment by mwcraig** [comment]
ipydatagrid is what I would have recommended -- I think that with the alternatives listed now this is good to go. May want to wait for @taldcroft though?

**Comment by MridulS** [comment]
@taldcroft just wanted to bump this up.

File: astropy/table/table.py
Comment (eerovaher) [inline astropy/table/table.py:1766]: Have you checked how this deprecation message actually looks like?

Comment (MridulS) [reply astropy/table/table.py:1766]: Looks decent-ish to me? (ipython)
Screenshot 2024-02-19 at 8 35 06 PM

(Jupyter lab interface)
Screenshot 2024-02-19 at 8 36 53 PM

Comment (eerovaher) [reply astropy/table/table.py:1766]: It looks to me like an extra space would improve alignment.
```suggestion
        message="""show_in_notebook() is deprecated as of 6.1 and to create
         interactive tables it is recommended to use dedicated tools like:
         - https://github.com/bloomberg/ipydatagrid
         - https://docs.bokeh.org/en/latest/docs/user_guide/interaction/widgets.html#datatable
         - https://dash.plotly.com/datatable""",
```

**Comment by eerovaher** [comment]
One more thing that we have failed to notice so far is that `show_in_notebook()` is mentioned in the `table` documentation: https://github.com/astropy/astropy/blob/55799978cd000e3d285bf3a8c10a91fb0e1032ac/docs/table/index.rst?plain=1#L112-L113
If we are deprecating the function then we shouldn't be advertising it in our documentation anymore.

**Comment by pllim** [comment]
Just in case, you could do a recursive grep on the whole repo checkout too, to smoke out the rest.

File: astropy/table/table.py
Comment (Cadair) [reply astropy/table/table.py]: I find show_in_browser to be very useful fwiw, although I don't really use the jsviewer flag that often.

**Review by eerovaher — COMMENTED** [review]
This should now be good enough to squash and merge, but I'm not a `table` maintainer.

**Review by taldcroft — APPROVED** [review]
Looks great, thanks @MridulS for getting this done!

**Comment by taldcroft** [comment]
And thanks to other reviewers for the inputs.

**Comment by taldcroft** [comment]
@MridulS - I've approved this, but can you track down the "show-in-notebook" issues that this fixes and put them in the description before we merge so they all get auto-closed?

**Comment by MridulS** [comment]
@taldcroft thanks! updated https://github.com/astropy/astropy/pull/15905#issue-2086728621 with the relevant issues.

**Comment by pllim** [comment]
So, is this good to merge now? 😸

**Comment by taldcroft** [comment]
> So, is this good to merge now? 😸

YES!

**Comment by pllim** [comment]
So happy to see one PR closes multiple issues. Thanks, all!

**Comment by pllim** [comment]
Looks like no everyone is happy with this, please see https://github.com/astropy/astropy/issues/16702 and advise. Thanks!

**Comment by bsipocz** [comment]
I run into this today as prepping notebooks for an upcoming AAS workshop. I see why it is removed, but otoh none of the offered alternatives are as nice our of the box experience, I mean I don't want to pull in all of bokeh just to have nicely formatted tables in a notebook. 

(we didn't use the `show_in_notebook` for anything fancy and interactive just to have a sensibly formatted, and not ellipsed table with nicely defaulted scroller and such. Part of the workshop was to showcase astropy, too and why Table is "better" than pandas (the question comes up regularly), but if any of these third-party table tools need to be pulled in we may just pull the plug and switch to pandas?).

**Comment by taldcroft** [comment]
> why Table is "better" than pandas (the question comes up regularly)

We should have a FAQ, but for the record (and for your tutorial), Table provides these capabilities that are not in pandas DataFrame:
- Robust support for table and column metadata. This is needed for representing data in file formats like FITS and VOTable.
- Support for N-dimensional column data, for example a table of thumbnail images in a source catalog.
- Support for columns of astropy Time, Quantity, and SkyCoord / Frame objects.
- General support for external "mixin" columns, for instance dask arrays.

**Comment by taldcroft** [comment]
As far as I can tell (from quick googling), pandas does not provide an interactive scrollable notebook interface for exploring notebooks. Google just shows 3rd party solutions. This choice by pandas would seem to support the astropy decision to delegate this kind of functionality to 3rd party packages.

>  I mean I don't want to pull in all of bokeh just to have nicely formatted tables in a notebook.

This sort of comment comes up regularly, and I respect your perspective but honestly do not understand it. I just installed bokeh and it required 6.9 Mb and completed in under a second, so I would just do it without a second thought and then use it.

**Comment by taldcroft** [comment]
There is this interesting news: https://blog.jupyter.org/ipydatagrid-is-now-part-of-project-jupyter-3b3dfb877664

I haven't looked into it carefully, but maybe we can leverage ipydatagrid and provide interactive Table display using this as the optional (?) dependency. BTW standalone ipydatagrid install is a little lighter at 3.1 Mb.

**Comment by pllim** [comment]
Thanks, all! Let's continue discussions of follow-up topics here:

* https://github.com/astropy/astropy/issues/16702
* https://github.com/astropy/astropy/issues/17146

**Comment by bsipocz** [comment]
If you forget interactivity (I don't necessarily need that, for more complex interactivity tasks we use firefly anyway), but if there was an easy opt in config list that would make Table printing look pretty in a notebook by default, I would be super happy with it. 
I know it shouldn't really be astropy's task to come up with this and keep maintaining it, but in practice, we already do something similar for visualization etc.

(I am thinking for my users who work in notebooks, and I really can't see it being a nice workflow to have them work with tables and any time they need to print more than a handful or rows I suggest them to pop into a third-party library instead. )

And my issue with adding bokeh just for this is not about size, but past experience of their changing API (it may very well be more stable now), and the technical debt I would introduce to this material just by introducing a new dependency with the sole purpose of pretty printing tables.

**Pull request merged**
